rust/src/graph/cycles.rs

```diff
@@ -0,0 +1,138 @@
+        .push(v.to_string());
+    }
+
+    fn find_cycles(&self) -> Vec<Vec<String>> {
+        let mut cycles = Vec::new();
+        let mut blocked = HashSet::new();
+        let mut stack = Vec::new();
+        let mut b_sets: HashMap<String, HashSet<String>> = HashMap::new();
+
+        for start in self.adj_list.keys() {
+            let mut subgraph = self.build_subgraph(start);
+            self.find_cycles_from(
+                start,
+                start,
+                &mut blocked,
+                &mut stack,
+                &mut b_sets,
+                &mut cycles,
+                &mut subgraph,
+            );
+        }
+
+        cycles
+    }
+
+    fn build_subgraph(&self, start: &String) -> HashMap<String, Vec<String>> {
+        self.adj_list
+            .iter()
+            .filter(|(node, _)| node >= &start)
+            .map(|(node, neighbors)| {
+                let filtered_neighbors: Vec<String> =
+                    neighbors.iter().filter(|n| n >= &start).cloned().collect();
+                (node.clone(), filtered_neighbors)
+            })
+            .collect()
+    }
+
+    fn find_cycles_from(
+        &self,
+        start: &String,
+        v: &String,
+        blocked: &mut HashSet<String>,
+        stack: &mut Vec<String>,
+        b_sets: &mut HashMap<String, HashSet<String>>,
+        cycles: &mut Vec<Vec<String>>,
+        subgraph: &mut HashMap<String, Vec<String>>,
+    ) -> bool {
+        let mut found_cycle = false;
+        stack.push(v.clone());
+        blocked.insert(v.clone());
+
+        let neighbors = subgraph.get(v).cloned().unwrap_or_default();
+        for w in neighbors {
+            if &w == start {
+                cycles.push(stack.clone());
+                found_cycle = true;
+            } else if !blocked.contains(&w) {
+                if self.find_cycles_from(start, &w, blocked, stack, b_sets, cycles, subgraph) {
+                    found_cycle = true;
+                }
+            }
+        }
+
+        if found_cycle {
+            self.unblock(v, blocked, b_sets);
+        } else {
+            if let Some(neighbors) = subgraph.get(v) {
+                for w in neighbors {
+                    b_sets.entry(w.clone()).or_default().insert(v.clone());
+                }
+            }
+        }
+
+        stack.pop();
+        found_cycle
+    }
+
+    fn unblock(&self, node: &String, blocked: &mut HashSet<String>, b_sets: &mut HashMap<String, HashSet<String>>) {
+        blocked.remove(node);
+        let mut stack = vec![node.clone()];
+
+        while let Some(n) = stack.pop() {
+            if let Some(dependent_nodes) = b_sets.remove(&n) {
+                for w in dependent_nodes {
+                    if blocked.remove(&w) {
+                        stack.push(w);
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl Graph {
+    pub fn find_cycles(
+        &self
+    ) -> Vec<Vec<String>> {
```

Comment:
🤔 I'm not sure we want to work with `String` here . Inside `Graph` I've designed the code to work in terms of `ModuleToken`s. We generally translate back to strings just for the python layer (i.e. in `GraphWrapper` in `lib.rs`).